src/indicator/algorithm/calculator/envelope.js

```diff
@@ -0,0 +1,36 @@
+import d3 from 'd3';
+
+export default function() {
+
+    var channel = 0.1;
+    var midValue = function(d) {
+        return d.value;
+    };
+    var envelope = function(data) {
+        data.forEach(function(s) {
+            if (midValue(s) !== undefined) {
```

Comment:
In order that calculators do not have to deal with leading 'undefined' values we have an adapter that takes care of that, resulting in simpler calculator implementations. See the following:

https://github.com/ScottLogic/d3fc/blob/915b2ff2d79096c8aa12cbac634038f62debcf4a/src/indicator/algorithm/calculator/macd.js#L18